### DISCUSSION TITLE
Should we remove the "experimental" tag from type-solving?

### DISCUSSION THREAD

**Discussion by BraisGabin** [detekt/detekt#4959] [General]
Right now a lot of our rules use type-solving and it is battle tested. It works! But we still say in the documentation that it is "experimental" but, to me, it is not experimental at all. It's one of the core features of detekt.

So I would like to remove the "experimental tag" from the type-solving, what do you think?

### Context
https://github.com/appKODE/detekt-rules-compose/issues/5#issuecomment-1155690360 The compose rule set doesn't use type-solving because it's tagged as experimental and I think that's really bad.

To me, the only thing that have some "unpulished edges" is the gradle integration. But I think that all of us agree that we are not going to add any breaking change there even with the "experimental" tag.

**Comment by cortinico** [comment]
Agree that is not experimental anymore.

On this topic, I should update our wording & task naming to something like:
- `detektLite` -> i.e. the current execution of `detekt` without Type Resolution
- `detektFull` -> i.e. Running with type resolution.

The Lite/Full naming will make easier for users to understand what is going on.

**Reply by BraisGabin** [reply]
I like those names!

So the idea would be:
- `detektLite` = `detekt`
- `detektFull` = `detektMain` && `detektTest`

**Reply by cortinico** [reply]
Essentially yes

**Reply by TWiStErRob** [reply]
I think `detektStatic` might be expressing more that there are no dynamic analysis going on.

**Reply by TWiStErRob** [reply]
Would making type resolution the default possible?
i.e.
 * `detekt`: `dependsOn(detektMain, detektTest, detektDebug, ...)`
 * `detektStatic`: doesn't care about compilations, just uses the module route as source set

Background: most developers don't care about these implementation details of Detekt (type resolution, configurations, multiple tasks, etc). They just want to "run Detekt", so `gradlew detekt` and fix the issues.

Also qualifying both full and lite versions just makes the users decide, would it be reasonable to have an encouraged default?

**Reply by cortinico** [reply]
> I think detektStatic might be expressing more that there are no dynamic analysis going on.

IMHO Static is confusing in this context as Detekt as a whole is always doing a static analysis (i.e. we're not running the code). The amount of available information for the analysis is different though.

**Reply by TWiStErRob** [reply]
oh, that's fair, `detekt` and `detektLite` then?

**Reply by cortinico** [reply]
I would stick to `detektLite`/`detektFull` and I would 'alias' `detekt` -> `detektFull`.

**Reply by TWiStErRob** [reply]
Aliasing breaks down with surprises around `tasks.named<Detekt>("detekt")`.

**Reply by BraisGabin** [reply]
I agree that detekt is the desired name for the task. But I don't think we should rename that until 2.0.

**Reply by TWiStErRob** [reply]
Little off topic: do you have a list of "well do it in 2.0" yet?

**Reply by BraisGabin** [reply]
Yes, we have, we moved to a discussion

**Reply by cortinico** [reply]
> Aliasing breaks down with surprises around `tasks.named<Detekt>("detekt")`.

IMHO this is expected and should be a breaking change. I.e. we'll probably change `detekt` to don't be a `Detekt` task anymore or so.

**Reply by TWiStErRob** [reply]
Related issue: https://github.com/detekt/detekt/issues/3122

**Reply by matejdro** [reply]
I think that having two tasks (`detektLite` / `detektFull`) is better than having a gradle property due to configuration cache. Changing the property would require gradle to reconfigure the build, while having two separate tasks can be cached normally.

**Reply by TWiStErRob** [reply]
@matejdro Aren't the "requested tasks" part of the configuration cache key? As far as I observed, when we ask for different tasks, the configuration cache is recalculated.

**Reply by matejdro** [reply]
The way I understand it, requested tasks are keys for the configuration caches, but properties are inputs.

So, in `detektLite`/`detektMain` scenario:

1. I run detekt without type solving (`detektLite`) - configuration cache for `detektLite` is computed and saved
2. I run detekt with type solving (`detektMain`) - configuration cache for `detektMain` is computed and saved
3. I run detekt without type solving (`detektLite`) again - configuration cache for `detektLite` is reused, we don't need to recompute the configuration

However, in custom property scenario:

1. I run detekt without type solving (`detekt -PenableTypeSolving=false`) - configuration cache for `detekt` is computed and saved
2. I run detekt with type solving (`detekt -PenableTypeSolving=true`) - input has changed, so configuration for task `detekt` needs to be re-computed
3. I run detekt without type solving (`detekt -PenableTypeSolving=false`) - input has changed, so configuration for task `detekt` needs to be re-computed

Re-running the task again (step 3) will only reuse configuration cache in `detektLite`/`detektMain` scenario

**Reply by TWiStErRob** [reply]
I see, thanks for the steps, it really helps understand what you meant and [from the docs](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:intro:how_does_it_work) I also read it similarly. You bring a good point about this behavior being problematic. However, I don't think this would be every-day user behavior, right? What would be the use case for running with type resolution on/off in sequence? I would imagine that projects would choose to accept type resolution as the default, or override it to be off, but not toggle between them often. (Except maybe in the initial phase when build engineers are trying to see differences in output, but that's the .01% of the use, the rest is CI and devs running it as configured.)

I made a [repro](https://github.com/TWiStErRob/repros/tree/main/detekt/typeResolution-configurationCacheFlag) to confirm this behavior and play around a bit; I also opened a [Gradle issue](https://github.com/gradle/gradle/issues/25039), because this looks strange. Properties invalidate the cache entries, but not used as part of the key. I'm curious why this is happening.

**Reply by matejdro** [reply]
Yeah, you are right, that this is an edge case. Just wanted to mention it in case there is a tie between the two options, as a minor point for the two tasks side.

**Reply by TWiStErRob** [reply]
I would like to note for maintainers that the internal implementation of the type resolution in DGP can be still be done with a flag, regardless of which user facing task pattern is chosen at the end:
https://github.com/TWiStErRob/repros/blob/main/detekt/typeResolution-configurationCacheFlag/build.gradle.kts
